BSP/Src/BSP.c

```diff
@@ -0,0 +1,187 @@
+    [BSP_GPIO_PA12] = {GPIO_AF5_SPI2},
+    [BSP_GPIO_PA13] = {}, //gpio/unused?
+    [BSP_GPIO_PA14] = {}, //gpio
+    [BSP_GPIO_PA15] = {}, //gpio
+    [BSP_GPIO_PB0] = {}, //disp7
+    [BSP_GPIO_PB1] = {}, //disp8
+    [BSP_GPIO_PB2] = {}, //disp9
+    [BSP_GPIO_PB3] = {GPIO_AF11_CAN3},
+    [BSP_GPIO_PB4] = {GPIO_AF11_CAN3},
+    [BSP_GPIO_PB5] = {GPIO_AF9_CAN2},
+    [BSP_GPIO_PB6] = {GPIO_AF9_CAN2},
+    [BSP_GPIO_PB7] = {}, //gpio
+    [BSP_GPIO_PB8] = {GPIO_AF4_I2C1},
+    [BSP_GPIO_PB9] = {GPIO_AF4_I2C1},
+    [BSP_GPIO_PB10] = {}, //disp10
+    [BSP_GPIO_PB11] = {},
+    [BSP_GPIO_PB12] = {GPIO_AF11_UART5},
+    [BSP_GPIO_PB13] = {GPIO_AF11_UART5},
+    [BSP_GPIO_PB14] = {},  //gpio
+    [BSP_GPIO_PB15] = {}, //gpio
+    [BSP_GPIO_PC0] = {}, //disp1
+    [BSP_GPIO_PC1] = {}, //disp2
+    [BSP_GPIO_PC2] = {}, //disp3
+    [BSP_GPIO_PC3] = {}, //disp4
+    [BSP_GPIO_PC4] = {}, //disp5
+    [BSP_GPIO_PC5] = {}, //disp6
+    [BSP_GPIO_PC6] = {}, //disp13
+    [BSP_GPIO_PC7] = {}, //disp14
+    [BSP_GPIO_PC8] = {}, //gpio
+    [BSP_GPIO_PC9] = {GPIO_AF4_I2C3}, 
+    [BSP_GPIO_PC10] = {GPIO_AF6_SPI3},
+    [BSP_GPIO_PC11] = {GPIO_AF6_SPI3},
+    [BSP_GPIO_PC12] = {GPIO_AF6_SPI3},
+    [BSP_GPIO_PC13] = {},
+    [BSP_GPIO_PD2] = {}
+};
+
+// Get HAL GPIO_PIN_XX from an enumerated pin
+uint8_t PinFromEnum(BSP_PINS pin) {
+    GPIO_TypeDef* pin_port = GET_GPIO(pin);
+    uint8_t final_pin_value;
+    if (pin_port == GPIOA) {
+        final_pin_value = (uint8_t)pin;
+    } else if (pin_port == GPIOB) {
+        final_pin_value = (uint8_t)pin - BSP_GPIO_PB0;
+    } else if (pin_port == GPIOC) {
+        final_pin_value = (uint8_t)pin - BSP_GPIO_PC0;
+    } else if (pin_port == GPIOD || pin == BSP_GPIO_PD2) {
+        final_pin_value = 2;
+    }
+    return final_pin_value;
+}
+
+// Determines (runtime) whether or not a pin AF mapping is valid
+bool isValidPinMapping(BSP_PINS pin, uint8_t mapped_function) {
+    for (uint8_t i = 0; i < MAX_PIN_AF; i++) {
+        if (VALID_AF_MAP[pin][i] == mapped_function) return true;
+    }
+    return false; // didn't match on any of the possible AFs for that pin, so bad mapping
+}
+
+/**
+ * @brief Configures the BSP layer according to the configuration specified in bsp_config.h
+ * 
+ * @details This function configures the BSP peripheral functions according to the alternate function
+ * mapping specified in bsp_config.h
+ * 
+ * The validity of your configuration is checked at runtime. Each pin's alternate function value is mapped
+ * into an array in memory, indexed by the pin, with a value of BSP_XXX alternate function value. This returned
+ * value is used as a secondary index into a LUT that translates BSP_XXX alternate function enum value into the 
+ * actual HAL alternate function value, allowing us to loop over the entire array of pins to configure their
+ * GPIO/AF properly, while only requiring modifications to the bsp_config.h file.
+ * 
+ */
+void BSP_Init() {
```

Comment:
Have to consider what happens to the MSP functions and if we could put some of the logic in there. Maybe we can make this function initialize init structs based on the config file and then allow the MSP functions to use those

Reply:
Half of the BSP files aren't using autogenerated MSP code however. I think BSP_Init should at the very least handle the GPIO setup because thats what 90% of the standard config functionality does. Actual peripheral init should be dispatched out to the corresponding BSP_XYZ file, but there's definitely a way to integrate the global handles with the BSP code.